### DISCUSSION TITLE
Can't get to REPL on M5STAMP-C3U (ESP32C3)

### DISCUSSION THREAD

**Discussion by thatbudakguy** [micropython/micropython#15850] [ESP32] [answered]
Hello,

I'm trying to access a REPL using `mpremote` on my M5STAMP C3U device. When I connect, I get the prompt:
```shell
$ mpremote connect /dev/cu.usbmodem2101

Connected to MicroPython at /dev/cu.usbmodem2101
Use Ctrl-] or Ctrl-x to exit this shell
```
...but then the shell hangs and there is no further output.

I'm on an M3 Mac (ARM) on macOS 14.6.1. I've compiled the ESP32 port of micropython v1.23 (with board set to ESP32_GENERIC_C3), which succeeded, and I was able to erase the flash and flash it:
```
$ python -m esptool --chip esp32c3 -p /dev/cu.usbmodem2101 -b 115200 --before default_reset --after hard_reset write_flash --flash_mode dio --flash_size 4MB --flash_freq 80m 0x0 build-ESP32_GENERIC_C3/bootloader/bootloader.bin 0x8000 build-ESP32_GENERIC_C3/partition_table/partition-table.bin 0x10000 build-ESP32_GENERIC_C3/micropython.bin

esptool.py v4.7.0
Serial port /dev/cu.usbmodem2101
Connecting....
Chip is ESP32-C3 (QFN32) (revision v0.3)
Features: WiFi, BLE, Embedded Flash 4MB (XMC)
Crystal is 40MHz
MAC: 68:67:25:82:3d:8c
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Flash will be erased from 0x00000000 to 0x00004fff...
Flash will be erased from 0x00008000 to 0x00008fff...
Flash will be erased from 0x00010000 to 0x001a9fff...
Compressed 17376 bytes to 11126...
Wrote 17376 bytes (11126 compressed) at 0x00000000 in 0.2 seconds (effective 567.7 kbit/s)...
Hash of data verified.
Compressed 3072 bytes to 115...
Wrote 3072 bytes (115 compressed) at 0x00008000 in 0.0 seconds (effective 632.9 kbit/s)...
Hash of data verified.
Compressed 1678480 bytes to 1050678...
Wrote 1678480 bytes (1050678 compressed) at 0x00010000 in 12.8 seconds (effective 1047.8 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
```

<s>Looking at [earlier discussion](https://github.com/orgs/micropython/discussions/12269Z), it seems that there used to be a separate ESP32C3-USB build with USB-UART support, but this was merged with the main ESP32_GENERIC_C3 build. I think I am in the b) category of the linked discussion (boards with REPL over serial and a serial/USB bridge on the board via /dev/ttyUSBx), but I'm not sure what else I need to do to get to the REPL over USB.</s>

Actually, I think what I need to do is enable the USB CDC at boot option, as [mentioned in the IDF docs](https://docs.espressif.com/projects/esp-idf/en/v5.0/esp32c3/api-guides/usb-serial-jtag-console.html#usb-serial-jtag-controller-console). But I don't know how to do this.

**Comment by davefes** [comment]
Just checking as I have only used mpremote a few times ... are you doing a Ctrl-x and then entering the `mpremote command you want` on the "command line"?

**Reply by thatbudakguy** [reply]
Not sure what you mean. My understanding is that `mpremote connect [device]` should give you a REPL. When I do it, after connecting I get no prompt and the shell does not respond to any input other than Ctrl-x, which disconnects me.

**Reply by davefes** [reply]
Not for me either.
Try:
`mpremote fs ls`

**Reply by thatbudakguy** [reply]
No output at all; it just hangs until Ctrl+C.

**Reply by davefes** [reply]
Hmmm.  Think there has been some discussion regarding MAC.  Have you got Linux running on something?

**Reply by thatbudakguy** [reply]
I don't, unfortunately.

**Comment by Jeff-Squared** [comment]
> does not respond to any input other than Ctrl-x

> No output at all; it just hangs until Ctrl+C

The second comment contradicts the first but it is interesting because I once had a device that had this behavior and ctrl c was the way I managed to get the **>>>** prompt

**Reply by thatbudakguy** [reply]
The behavior is actually different for the two commands.
- `mpremote connect` prints the "connected" message and will respond to Ctrl-x, but not Ctrl-C.
- `mpremote fs ls` prints nothing and will only respond to Ctrl-C.

**Reply by davefes** [reply]
`mpremote connect auto`

**Reply by thatbudakguy** [reply]
That's the same as `mpremote connect`; it detects and connects to `/dev/cu.usbmodem2101` which is my device.

**Reply by davefes** [reply]
I think you need to apply `divide and conquer`.

Have you got a ESP32-WROOM (ESP32-Generic)?

**Reply by thatbudakguy** [reply]
No, but I've got an M5STACK TOUGH handy, which is a regular ESP32. I compiled an ESP32_GENERIC build, but it seems like the same kind of thing is happening...all `python -m esptool` commands to flash/erase just hang at:
```
esptool.py v4.7.0
Serial port /dev/tty.usbserial-54D80277501
```
No more output appears, and I can't even Ctrl-C to quit...I have to unplug the device or quit my terminal to get it to relinquish the shell.

This device has a  CH9102 USB-TTL chip in it, so different from the ESP32C3, but I seem to be facing the same problem.

**Reply by davefes** [reply]
Hmmm, my ESP32 have CP2102 ... if they are similar then I think you probably need to look harder for issues with MAC OS.

Maybe:
https://github.com/orgs/micropython/discussions/13545

Maybe look around the old forum:
https://forum.micropython.org/search.php?search_id=unreadposts

**Comment by thatbudakguy** [comment]
This seems related to https://github.com/micropython/micropython/commit/0a11832cddbeed57bf0267afa0c833584817312d. I tried using the most recent `ESP32_GENERIC_C3/sdkconfig.c3usb`, but no luck. Not sure what config options to enable/disable.

**Reply by davefes** [reply]
I looked in the latest source file `/ports/esp32/usb.c` and I see the change, so I would expect that the latest build would have it.  Have you tried a preview?

**Reply by thatbudakguy** [reply]
What I ended up doing was trying an older version. [This repo](https://github.com/gandro/micropython-m5stamp-c3u) suggested v1.19.1, so I went and downloaded a prebuilt binary and flashed it. This worked! After flashing, when I did `mpremote connect auto`, I got a REPL (after hitting `enter`).

So, something is different about v1.19.1 – I think it's these settings: https://github.com/micropython/micropython/blob/9b486340da22931cde82872f79e1c34db959548b/ports/esp32/boards/GENERIC_C3_USB/sdkconfig.board#L7-L9

**Reply by robert-hh** [reply]
If I see the schematics right, the MpSTAMP-C3U need REPL on UART0. 
I have different ESP32C3 boards with genuine USB and USB over Serial. Both work with the ESP32 Generic firmware and a firmware built from the actual master.

**Reply by thatbudakguy** [reply, marked as answer]
I downloaded the [v1.24.0-preview.286.g042693496](https://micropython.org/resources/firmware/ESP32_GENERIC_C3-20240904-v1.24.0-preview.286.g042693496.bin) image and that worked too! Now I'm kind of confused. But in any case, the problem is solved...maybe I just need to use prebuilt firmware images instead of building from source from now on.
